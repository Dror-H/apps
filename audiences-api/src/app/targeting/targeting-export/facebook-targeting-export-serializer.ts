import {
  AdAccountDTO,
  AdSetCreationDTO,
  CampaignCreationDTO,
  CampaignEffectiveStatusType,
  CampaignStatusType,
  FacebookBidStrategyEnum,
  FacebookCampaignCreationDto,
  PacingType,
  SupportedProviders,
  TargetingDto,
} from '@instigo-app/data-transfer-object';
import { format } from 'date-fns';

export class FacebookTargetingExportSerializer {
  public constructor(private name: string, private _adAccount: AdAccountDTO, private _targeting: TargetingDto) {}

  public get adAccount(): AdAccountDTO {
    return this._adAccount;
  }

  public get targeting(): TargetingDto {
    return this._targeting;
  }

  public serializeCampaignObject(): CampaignCreationDTO<FacebookCampaignCreationDto> {
    return {
      name: `Instigo Audiences | ${this.name || this.getFormattedDate()}`,
      status: CampaignStatusType.PAUSED,
      effectiveStatus: CampaignEffectiveStatusType.IN_PROCESS,
      provider: SupportedProviders.FACEBOOK,
      adAccount: this.adAccount,
      providerSpecificFields: {
        objective: 'LINK_CLICKS',
        specialAdCategories: null,
        buyingType: 'auction',
        spendCap: 20,
        dailyBudget: 20,
        bidStrategy: FacebookBidStrategyEnum.LOWEST_COST_WITHOUT_CAP,
      },
    } as CampaignCreationDTO<FacebookCampaignCreationDto>;
  }

  public serializeAdSetCreationObject(): Partial<AdSetCreationDTO> {
    return {
      name: `Instigo Audiences | ${this.name || this.getFormattedDate()}`,
      status: CampaignStatusType.ACTIVE,
      optimizationGoal: 'LINK_CLICKS',
      billingEvent: 'IMPRESSIONS',
      bidStrategy: FacebookBidStrategyEnum.LOWEST_COST_WITH_BID_CAP,
      bidAmount: null,
      provider: SupportedProviders.FACEBOOK,
      targeting: this.targeting,
      startTime: new Date(Date.now()),
      pacingType: [PacingType.STANDARD],
      adAccount: this.adAccount,
      campaign: null,
    };
  }

  private getFormattedDate(): string {
    return format(new Date(), 'MMM dd Y HH:mm:ss');
  }
}
