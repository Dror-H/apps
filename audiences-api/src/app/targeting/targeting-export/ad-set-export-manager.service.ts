import { ForbiddenException, Inject, Injectable, Logger } from '@nestjs/common';
import { FacebookTargetingExportSerializer } from '@audiences-api/targeting/targeting-export/facebook-targeting-export-serializer';
import {
  AdAccountDTO,
  AdSetDTO,
  CampaignDTO,
  CampaignStatusType,
  SupportedProviders,
} from '@instigo-app/data-transfer-object';
import { PrismaService } from '@audiences-api/prisma/prisma.service';
import { ThirdPartyAdSetApiService, ThirdPartyFacebookExportApiService } from '@instigo-app/third-party-connector';

@Injectable()
export class AdSetExportManagerService {
  private readonly logger = new Logger(AdSetExportManagerService.name);

  @Inject(PrismaService)
  private readonly prismaService: PrismaService;

  @Inject(ThirdPartyFacebookExportApiService)
  private readonly thirdPartyService: ThirdPartyFacebookExportApiService;

  @Inject(ThirdPartyAdSetApiService)
  private readonly thirdPartyAdSetApiService: ThirdPartyAdSetApiService;

  public async updateAdSetTargeting(
    manager: FacebookTargetingExportSerializer,
    accessToken: string,
    campaign: CampaignDTO,
  ): Promise<AdSetDTO> {
    try {
      const adSetId = await this.getAdSetId(manager.adAccount.id);
      if (adSetId) {
        const facebookAdSet = await this.getFacebookAdSetById(adSetId, accessToken);
        facebookAdSet.targeting = manager.targeting;
        if (this.isUsableAdSet(facebookAdSet)) {
          await this.updateAdSet(accessToken, facebookAdSet);
          return facebookAdSet;
        }
      }
      return await this.createAdSetIfDoesntExist(manager, accessToken, campaign);
    } catch (err) {
      this.logger.error(err);
      throw new ForbiddenException(err);
    }
  }

  private isUsableAdSet(adSet: AdSetDTO): boolean {
    return !(adSet.status == CampaignStatusType.DELETED || adSet.status == CampaignStatusType.ARCHIVED);
  }

  private async getAdSetId(adAccountId: string): Promise<string> {
    const { adsetId } = await this.prismaService.adAccount.findFirst({
      where: { id: adAccountId },
      select: { adsetId: true },
    });
    return adsetId;
  }

  private async saveAdSetIdForAdAccount(adAccount: AdAccountDTO, adSetId: string): Promise<string> {
    return (
      await this.prismaService.adAccount.update({
        where: {
          id: adAccount.id,
        },
        data: {
          adsetId: adSetId,
        },
      })
    ).adsetId;
  }

  private async getFacebookAdSetById(adSetId: string, accessToken: string): Promise<AdSetDTO> {
    return await this.thirdPartyService.getAdSetById({
      accessToken,
      adSetId: adSetId,
    });
  }

  private async createAdSetIfDoesntExist(
    manager: FacebookTargetingExportSerializer,
    accessToken: string,
    campaign: CampaignDTO,
  ): Promise<AdSetDTO> {
    const adSetDTO = manager.serializeAdSetCreationObject();
    adSetDTO.campaign = campaign;
    const newAdSet = await this.thirdPartyService.createAdSet({
      accessToken,
      adAccountProviderId: manager.adAccount.id,
      adSet: adSetDTO,
    });
    await this.saveAdSetIdForAdAccount(manager.adAccount, newAdSet.providerId);
    return newAdSet;
  }

  private updateAdSet(accessToken: string, adSet: AdSetDTO): Promise<AdSetDTO> {
    return this.thirdPartyAdSetApiService.bulkPatch({
      provider: SupportedProviders.FACEBOOK,
      accessToken,
      adSets: [adSet],
    });
  }
}
