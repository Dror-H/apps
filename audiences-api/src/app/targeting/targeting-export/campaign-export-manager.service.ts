import { ForbiddenException, Inject, Injectable, Logger } from '@nestjs/common';
import { PrismaService } from '@audiences-api/prisma/prisma.service';
import { FacebookTargetingExportSerializer } from '@audiences-api/targeting/targeting-export/facebook-targeting-export-serializer';
import { AdAccountDTO, CampaignDTO, CampaignStatusType } from '@instigo-app/data-transfer-object';
import { ThirdPartyFacebookExportApiService } from '@instigo-app/third-party-connector';

@Injectable()
export class CampaignExportManagerService {
  private readonly logger = new Logger(CampaignExportManagerService.name);

  @Inject(PrismaService)
  private readonly prismaService: PrismaService;

  @Inject(ThirdPartyFacebookExportApiService)
  private readonly thirdPartyService: ThirdPartyFacebookExportApiService;

  public async getCampaign(manager: FacebookTargetingExportSerializer, accessToken: string): Promise<CampaignDTO> {
    try {
      const campaignId = await this.getCampaignId(manager.adAccount.id);
      if (campaignId) {
        const facebookCampaign = await this.getFacebookCampaignById(campaignId, accessToken);
        if (this.isUsableCampaign(facebookCampaign)) {
          return facebookCampaign;
        }
      }
      return await this.createCampaignIfDoesntExist(manager, accessToken);
    } catch (err) {
      this.logger.error(err);
      throw new ForbiddenException(err);
    }
  }

  private async getCampaignId(adAccountId: string): Promise<string> {
    const { campaignId } = await this.prismaService.adAccount.findFirst({
      where: { id: adAccountId },
      select: { campaignId: true },
    });
    return campaignId;
  }

  private async saveCampaignIdForAdAccount(adAccount: AdAccountDTO, campaignId: string): Promise<string> {
    return (
      await this.prismaService.adAccount.update({
        where: {
          id: adAccount.id,
        },
        data: {
          campaignId,
        },
      })
    ).campaignId;
  }

  private isUsableCampaign(campaign: CampaignDTO): boolean {
    return !(campaign.status !== CampaignStatusType.PAUSED || campaign.adSetsCount >= 50);
  }

  private async getFacebookCampaignById(campaignId: string, accessToken: string): Promise<CampaignDTO> {
    return await this.thirdPartyService.getCampaignById({
      accessToken,
      campaignId,
    });
  }

  private async createCampaignIfDoesntExist(
    manager: FacebookTargetingExportSerializer,
    accessToken: string,
  ): Promise<CampaignDTO> {
    const campaignDTO = manager.serializeCampaignObject();
    const newCampaign = await this.thirdPartyService.createCampaign({
      accessToken,
      adAccountProviderId: manager.adAccount.id,
      campaign: campaignDTO,
    });
    await this.saveCampaignIdForAdAccount(manager.adAccount, newCampaign.providerId);
    return newCampaign;
  }
}
