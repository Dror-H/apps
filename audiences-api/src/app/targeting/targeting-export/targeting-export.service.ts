import { Inject, Injectable, Logger } from '@nestjs/common';
import { AdAccountDTO, AdSetDTO, CampaignDTO, TargetingDto } from '@instigo-app/data-transfer-object';
import { FacebookTargetingExportSerializer } from '@audiences-api/targeting/targeting-export/facebook-targeting-export-serializer';
import { CampaignExportManagerService } from '@audiences-api/targeting/targeting-export/campaign-export-manager.service';
import { AdSetExportManagerService } from '@audiences-api/targeting/targeting-export/ad-set-export-manager.service';

@Injectable()
export class TargetingExportService {
  private readonly logger = new Logger(TargetingExportService.name);

  @Inject(CampaignExportManagerService)
  private readonly campaignExportManager: CampaignExportManagerService;

  @Inject(AdSetExportManagerService)
  private readonly adsetExportManager: AdSetExportManagerService;

  public async facebookExport(
    targeting: TargetingDto,
    adAccount: AdAccountDTO,
    accessToken: string,
    name: string,
  ): Promise<{ campaign: CampaignDTO; adSet: Partial<AdSetDTO> }> {
    this.logger.log('Facebook export started');
    const manager = new FacebookTargetingExportSerializer(name, adAccount, targeting);
    const campaign = await this.campaignExportManager.getCampaign(manager, accessToken);
    const adSet = await this.adsetExportManager.updateAdSetTargeting(manager, accessToken, campaign);
    return { campaign, adSet };
  }
}
