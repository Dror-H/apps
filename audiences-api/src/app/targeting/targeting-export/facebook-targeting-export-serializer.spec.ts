import { FacebookTargetingExportSerializer } from '@audiences-api/targeting/targeting-export/facebook-targeting-export-serializer';
import { AdAccountDTO, SupportedProviders, TargetingDto } from '@instigo-app/data-transfer-object';

const adAccount = {
  providerId: 'providerId',
  name: 'test-ad account',
  status: 'ACTIVE',
  currency: 'EUR',
  businessId: 'businessId',
  businessName: 'some business name',
  businessProfilePicture: 'string',
  provider: SupportedProviders.FACEBOOK,
  businessCity: 'string',
  businessStreet: 'string',
  businessStreet2: 'string',
  businessZip: 'string',
  businessCountryCode: 'string',
  timezoneId: 'string',
  timezoneName: 'string',
  timezoneOffsetHoursUtc: 'string',
  accessToken: 'string',
  totalAmountSpent: 'string',
  disableReason: 'string',
  adAccountStatusDescription: 'any',
} as AdAccountDTO;

describe('FacebookTargetingExportSerializer', () => {
  it('should serialize the campaign', () => {
    const campaign = new FacebookTargetingExportSerializer(
      'Audience Name',
      adAccount,
      {} as TargetingDto,
    ).serializeCampaignObject();
    expect(campaign).toMatchSnapshot();
  });

  it('should create the adSet', () => {
    jest.useFakeTimers().setSystemTime(new Date('2020-01-01'));

    const adSet = new FacebookTargetingExportSerializer(
      'Audience Name',
      adAccount,
      {} as TargetingDto,
    ).serializeAdSetCreationObject();
    expect(adSet).toMatchSnapshot();
  });
});
