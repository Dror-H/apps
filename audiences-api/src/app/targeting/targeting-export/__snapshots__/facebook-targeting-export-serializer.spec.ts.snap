// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FacebookTargetingExportSerializer should create the adSet 1`] = `
Object {
  "adAccount": Object {
    "accessToken": "string",
    "adAccountStatusDescription": "any",
    "businessCity": "string",
    "businessCountryCode": "string",
    "businessId": "businessId",
    "businessName": "some business name",
    "businessProfilePicture": "string",
    "businessStreet": "string",
    "businessStreet2": "string",
    "businessZip": "string",
    "currency": "EUR",
    "disableReason": "string",
    "name": "test-ad account",
    "provider": "facebook",
    "providerId": "providerId",
    "status": "ACTIVE",
    "timezoneId": "string",
    "timezoneName": "string",
    "timezoneOffsetHoursUtc": "string",
    "totalAmountSpent": "string",
  },
  "bidAmount": null,
  "bidStrategy": "LOWEST_COST_WITH_BID_CAP",
  "billingEvent": "IMPRESSIONS",
  "campaign": null,
  "name": "Instigo Audiences | Audience Name",
  "optimizationGoal": "LINK_CLICKS",
  "pacingType": Array [
    "standard",
  ],
  "provider": "facebook",
  "startTime": 2020-01-01T00:00:00.000Z,
  "status": "ACTIVE",
  "targeting": Object {},
}
`;

exports[`FacebookTargetingExportSerializer should serialize the campaign 1`] = `
Object {
  "adAccount": Object {
    "accessToken": "string",
    "adAccountStatusDescription": "any",
    "businessCity": "string",
    "businessCountryCode": "string",
    "businessId": "businessId",
    "businessName": "some business name",
    "businessProfilePicture": "string",
    "businessStreet": "string",
    "businessStreet2": "string",
    "businessZip": "string",
    "currency": "EUR",
    "disableReason": "string",
    "name": "test-ad account",
    "provider": "facebook",
    "providerId": "providerId",
    "status": "ACTIVE",
    "timezoneId": "string",
    "timezoneName": "string",
    "timezoneOffsetHoursUtc": "string",
    "totalAmountSpent": "string",
  },
  "effectiveStatus": "IN_PROCESS",
  "name": "Instigo Audiences | Audience Name",
  "provider": "facebook",
  "providerSpecificFields": Object {
    "bidStrategy": "LOWEST_COST_WITHOUT_CAP",
    "buyingType": "auction",
    "dailyBudget": 20,
    "objective": "LINK_CLICKS",
    "specialAdCategories": null,
    "spendCap": 20,
  },
  "status": "PAUSED",
}
`;
